quinn-proto/src/connection/streams.rs

```diff
@@ -1201,6 +1206,13 @@ pub enum StreamEvent {
         /// Error code supplied by the peer if the stream was stopped
         stop_reason: Option<VarInt>,
     },
+    /// A STOP_SENDING have been recieved for an outgoing stream
+    Stopped {
```

Comment:
This is already exposed via `WriteError::Stopped` and `FinishError::Stopped`. Is there a case those don't cover?

Reply:
This is exactly what thought at first. But Hyper wants to poll for STOP_SENDING before a response as been started. I guess this helps in heavy load conditions, where you want to handle clients abandon the lightest way as possible.

Reply:
What constitutes starting a response?

Reply:
I'm still a little confused. My starting point for this PR is [the `poll_reset()` h2 method](https://docs.rs/h2/0.2.5/h2/struct.SendStream.html#method.poll_reset), which belongs to the equivalent of `SendData` in quinn-h3. It seems intended to be polled before issuing some data to send. I think it should be polled while the web application is processing the request. The request has been passed to the upper layer but no data is ready for sending. At this time, the client can timeout, or react to a new action and cancel loading a resource or whatever (think of someone smashing a `next` link for example).

Reply:
That sure sounds like a synchronous check, i.e. `fn is_reset(&self) -> bool`, should be sufficient.

Reply:
Maybe I don't fully understand the use case then. It seems to me doing that would turn this behavior into active polling. Task waking is still useful here because we want to receive stop events for many requests, while serving many other requests.

Reply:
I think I misunderstood your example. You're saying that you'd select on stop and some asynchronous work required to prepare a response?

Reply:
Essentially, an HTTP lib receives a request, pass it to the application. At this point a computation is done that can take time (DB access, heavy load etc) and will result into a response.

While the application prepares this, the lib as one role: listen for cancellation events from the peer, and pass it to the application as soon as possible so the expensive computation of preparing a response can be shut down. 

At larger scale, a server with heavy load will want to cleanup cancelled requests as fast as possible to recover resources as they aren't needed anymore. Especially when the clients start to timeout because of the load, or people start smashing buttons, because of the load.

When the response is ready, you can totally pass the `STOP_SENDING` event to the app by throwing errors through any API that is polling for transmit. But until the response is ready for writing, this API stays untouched. 

This is why the App task responsible of the request processing needs to be waken instantly when receiving such a cancellation event.

But again, that only is a humble comprehension of this API needed by hyper.

Reply:
Okay, that seems reasonable; let's expose a future. I don't have a lot of opinions about the structure of HTTP things; my focus is more on the consequences at the transport layer.